### PR TITLE
jobs,*: read job payload and progress from the job_info table

### LINKED ISSUE #97762: jobs: migrate jobs to write their Payload and Progress to the jobs_info table
In https://github.com/cockroachdb/cockroach/pull/84866 we introduced a `system.jobs_info` table that is to be used to store job-specific state. This table allows a job to access and update information about its execution without adversely impacting the execution of other jobs. This is explained in detail in https://github.com/cockroachdb/cockroach/pull/82638.

As a first step, we want to teach jobs to write their Payload and Progress in their current form but to the jobs_info table instead of the jobs table. This migration will roughly involve:

1. Double writes for the Payload and Progress of jobs to both the jobs_info and jobs table.

2. A migration to backfill job_info with the Payload and Progress of all job records in the system.jobs table.

3. A change to read methods in the registry to pull the Payload and Progress from the job_info table instead of the jobs table.

In the future, jobs may want to decompose their Payload and Progress protos further and write different fields as their own rows in the jobs_info table. This is outside the scope of this issue.

Jira issue: CRDB-24874

Epic CRDB-8964

### PR DESCRIPTION
**Pull request by adityamaru**
This change touches all the parts of the code that were
relying on the `system.jobs` table to read the payload and/or
progress corresponding to a job. If the cluster version has advanced
past V23_1JobInfoTableIsBackfilled, every job in the system.jobs table
has a payload and progress entry written to the `job_info` table.
This change migrates callers to use this new table when reading the
payload and progress of a job.

The most important changes are in the logic that drives `crdb_internal.system_jobs`,
which in turn drives `crdb_internal.jobs` and `SHOW JOBS`. Additionally,
there are changes in how the registry resolves the progress and payload
when loading or resuming a job.

Several tests that read the payload and progress from `system.jobs` now
rely on `crdb_internal.system_jobs` to fetch this information from either
the jobs table or job_info table depending on the cluster version.

Fixes: https://github.com/cockroachdb/cockroach/issues/97762

Release note: None

### CODE REVIEW

File: pkg/ccl/changefeedccl/changefeed_test.go [hunk 1]
```diff
@@ -733,7 +733,8 @@ func TestChangefeedCursor(t *testing.T) {
 		// statement timestamp, so only verify this for enterprise.
 		if e, ok := fooLogical.(cdctest.EnterpriseTestFeed); ok {
 			var bytes []byte
-			sqlDB.QueryRow(t, `SELECT payload FROM system.jobs WHERE id=$1`, e.JobID()).Scan(&bytes)
+			sqlDB.QueryRow(t, `SELECT payload FROM "".crdb_internal.system_jobs WHERE id = $1`,
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/changefeed_test.go:736]: curious: why do we need "" ?

File: pkg/jobs/jobs.go [hunk 2]
```diff
@@ -750,6 +752,78 @@ func HasJobNotFoundError(err error) bool {
...
+		}
+		if err := protoutil.Unmarshal(payloadBytes, payload); err != nil {
+			return nil, nil, err
+		}
+
+		progressBytes, exists, err := infoStorage.GetLegacyProgress(ctx)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "failed to get payload for job %d", j.ID())
+		}
+		if !exists {
```
Comment (miretskiy) [inline pkg/jobs/jobs.go:782]: do we intend on making progress optional at some point?

File: pkg/jobs/adopt.go [hunk 3]
```diff
@@ -144,10 +146,19 @@ COALESCE(last_run, created) + least(
...
-	resumeQueryWhereBase   = `id = $1 AND claim_session_id = $2`
+	// resumeQueryBaseCols selects NULL values for the payload and progress that
+	// will be read from the system.job_info table. This allows us to get results
+	// aligned with deprecatedResumeQueryBaseCols below.
+	resumeQueryBaseCols    = "status, NULL, NULL, crdb_internal.sql_liveness_is_alive(claim_session_id)"
 	resumeQueryWithBackoff = `SELECT ` + resumeQueryBaseCols + `, ` + canRunClause + ` AS can_run,` +
 		` created_by_type, created_by_id  FROM system.jobs, ` + canRunArgs + " WHERE " + resumeQueryWhereBase
+
+	// deprecatedResumeQueryBaseCols loads the payload and progress from
+	// system.jobs instead of the system.job_info table.
```
Comment (miretskiy) [inline pkg/jobs/adopt.go:157]: can you add a todo or some such to deprecated this after release XXX?

**Review by miretskiy — APPROVED** [review]
I really only have nits.

File: pkg/ccl/changefeedccl/changefeed_test.go (hunk 1 above)
Comment (adityamaru) [reply pkg/ccl/changefeedccl/changefeed_test.go]: Added a comment to `jobs_verification.go` and changed most test places to use that query instead.

File: pkg/jobs/jobs.go (hunk 2 above)
Comment (adityamaru) [reply pkg/jobs/jobs.go:782]: _maybe_. On job creation we expect the job record to have a valid, non-null progress and payload so unless the user manually deletes the progress for a job from `system.job_info` we will always have an entry. That said `SHOW JOBS` goes out of its way to handle the absence of a progress row for a job. So I dunno we should decide if having progress is mandatory or not and then apply that through the system.

**Review by dt — APPROVED** [review]
Nice work!

**Comment by adityamaru** [comment]
`TestDecommission` is a flake - https://github.com/cockroachdb/cockroach/issues/96630
The other stress failure is during test server startup: `optInToDiagnosticsStatReporting: setting updated but timed out waiting to read new value` so nothing seems wrong with the test per-se. Will investigate separately.

**Comment by adityamaru** [comment]
bors retry

**Pull request merged**